src/provider/states/create_config.rs

```diff
@@ -0,0 +1,296 @@
+                    }
+                }
+            }
+        }
+        get_config_maps
+    }
+
+    async fn retrieve_config_map(
+        &self,
+        client: Client,
+        name: String,
+    ) -> Result<ConfigMap, StackableError> {
+        let config_maps: Api<ConfigMap> = Api::namespaced(client.clone(), "default");
+
+        Ok(config_maps.get(&name).await?)
+    }
+
+    fn apply_config_map(
+        &self,
+        map: ConfigMap,
+        target_directory: PathBuf,
+        template_data: &BTreeMap<String, String>,
+    ) -> Result<(), StackableError> {
+        let config_map_name = map.metadata.name.unwrap_or(String::from("undefined"));
+        debug!(
+            "applying configmap {} to directory {:?}",
+            &config_map_name, target_directory
+        );
+        if !(&target_directory.is_dir()) {
+            info!("creating config directory {:?}", target_directory);
+            fs::create_dir_all(&target_directory)?;
+        }
+        if let Some(data) = map.data {
+            debug!("Map contained keys: {:?}", &data.keys());
+            for key in data.keys() {
+                debug!("found key: {} in configmap {}", key, &config_map_name);
+                if let Some(content) = data.get(key) {
+                    trace!("content of key: {}", &content);
+                    debug!("rendering");
+                    let rendered_content = CreatingConfig::render_config_template(
+                        template_data.clone(),
+                        content.clone(),
+                    )?;
+                    debug!("done rendering");
+                    let target_file = target_directory.join(&key);
+
+                    // TODO: compare existing file with intended state
+                    if CreatingConfig::needs_update(&target_file, &rendered_content)? {
+                        debug!(
+                            "writing content of map entry {} to file {:?}",
+                            key, target_file
+                        );
+                        let write_result = fs::write(target_directory.join(&key), rendered_content);
+                        match write_result {
+                            Ok(()) => debug!("write of file {:?} successful!", target_file),
+                            Err(e) => error!("write of file {:?} failed: {}", target_file, e),
+                        }
+                    } else {
+                        debug!("No update needed for {:?}", target_file);
+                    }
+                } else {
+                    info!("No content found for key {}", key);
+                }
+            }
+        } else {
+            debug!("No data found in ConfigMap..");
+        }
+        Ok(())
+    }
+
+    fn needs_update(target_file: &PathBuf, content: &str) -> Result<bool, StackableError> {
+        if target_file.is_file() {
+            let current_content = read_to_string(target_file)?;
```

Comment:
Probably fine for now but should we ever deal with large files we might want to work on checksums instead and not read the files into memory

Reply:
We can look at that, sure. For most of the initial tools on our roadmap though I can't think of one with large enough config files to spend the time now.